cpp/src/arrow/util/compression_lz4.cc

```diff
@@ -225,20 +231,90 @@ class LZ4Compressor : public Compressor {
 };
 
 // ----------------------------------------------------------------------
-// Lz4 codec implementation
+// Lz4 frame codec implementation
 
-Result<std::shared_ptr<Compressor>> Lz4Codec::MakeCompressor() {
+struct Lz4FrameCodec::Impl {
+  Impl() : prefs_(DefaultPreferences()) {}
+
+  LZ4F_preferences_t prefs_;
+};
+
+Lz4FrameCodec::Lz4FrameCodec() : impl_(new Impl()) {}
+
+Lz4FrameCodec::~Lz4FrameCodec() {}
+
+Result<std::shared_ptr<Compressor>> Lz4FrameCodec::MakeCompressor() {
   auto ptr = std::make_shared<LZ4Compressor>();
   RETURN_NOT_OK(ptr->Init());
   return ptr;
 }
 
-Result<std::shared_ptr<Decompressor>> Lz4Codec::MakeDecompressor() {
+Result<std::shared_ptr<Decompressor>> Lz4FrameCodec::MakeDecompressor() {
   auto ptr = std::make_shared<LZ4Decompressor>();
   RETURN_NOT_OK(ptr->Init());
   return ptr;
 }
 
+int64_t Lz4FrameCodec::MaxCompressedLen(int64_t input_len,
+                                        const uint8_t* ARROW_ARG_UNUSED(input)) {
+  return static_cast<int64_t>(
+      LZ4F_compressFrameBound(static_cast<size_t>(input_len), &impl_->prefs_));
+}
+
+Result<int64_t> Lz4FrameCodec::Compress(int64_t input_len, const uint8_t* input,
+                                        int64_t output_buffer_len,
+                                        uint8_t* output_buffer) {
+  auto output_len =
+      LZ4F_compressFrame(output_buffer, static_cast<size_t>(output_buffer_len), input,
+                         static_cast<size_t>(input_len), &impl_->prefs_);
+  if (LZ4F_isError(output_len)) {
+    return LZ4Error(output_len, "Lz4 compression failure: ");
+  }
+  return static_cast<int64_t>(output_len);
+}
+
+Result<int64_t> Lz4FrameCodec::Decompress(int64_t input_len, const uint8_t* input,
+                                          int64_t output_buffer_len,
+                                          uint8_t* output_buffer) {
+  ARROW_ASSIGN_OR_RAISE(auto decomp, MakeDecompressor());
```

Comment:
I assume that this object is stateful and therefore must be created at each invocation of `Decompress`?

Reply:
Yes. It may be possible to avoid the allocation by inlining the decompress loop, but I'm not sure that would gain much.